File: razer-sdk/src/test/java/com/bonepl/razersdk/color/TransitionColorTest.java
package com.bonepl.razersdk.color;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class TransitionColorTest {

    @Test
    void testTooFewStepsTransition() {
        assertThrows(IllegalArgumentException.class, () -> new TransitionColor(StaticColor.RED, StaticColor.GREEN, -1));
        assertThrows(IllegalArgumentException.class, () -> new TransitionColor(StaticColor.RED, StaticColor.GREEN, 0));
        assertThrows(IllegalArgumentException.class, () -> new TransitionColor(StaticColor.RED, StaticColor.GREEN, 1));
    }

    @Test
    void testTwoStepTransition() {
        TransitionColor transitionColor = new TransitionColor(StaticColor.RED, StaticColor.GREEN, 2);
        assertEquals(StaticColor.RED, transitionColor.getColor());
        assertFalse(transitionColor.transitionFinished());
        assertEquals(StaticColor.GREEN, transitionColor.getColor());
        assertTrue(transitionColor.transitionFinished());
    }

    @Test
    void testThreeStepTransition() {
        TransitionColor transitionColor = new TransitionColor(StaticColor.RED, StaticColor.GREEN, 3);
        assertEquals(StaticColor.RED, transitionColor.getColor());
        assertFalse(transitionColor.transitionFinished());
        assertEquals(new StaticColor(128,127,0), transitionColor.getColor());
        assertFalse(transitionColor.transitionFinished());
        assertEquals(StaticColor.GREEN, transitionColor.getColor());
        assertTrue(transitionColor.transitionFinished());
    }
}

File: test/razer-sdk/color/TransitionColor.test.ts
import {describe, expect, test} from 'vitest';
import {TransitionColor} from '../../../src/razer-sdk/color/TransitionColor.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';

describe('TransitionColorTest', () => {
    test('testTooFewStepsTransition', () => {
        expect(() => new TransitionColor(StaticColor.RED, StaticColor.GREEN, -1)).toThrow();
        expect(() => new TransitionColor(StaticColor.RED, StaticColor.GREEN, 0)).toThrow();
        expect(() => new TransitionColor(StaticColor.RED, StaticColor.GREEN, 1)).toThrow();
    });

    test('testTwoStepTransition', () => {
        const transitionColor = new TransitionColor(StaticColor.RED, StaticColor.GREEN, 2);
        expect(transitionColor.getColor()).toEqual(StaticColor.RED);
        expect(transitionColor.transitionFinished()).toBe(false);
        expect(transitionColor.getColor()).toEqual(StaticColor.GREEN);
        expect(transitionColor.transitionFinished()).toBe(true);
    });

    test('testThreeStepTransition', () => {
        const transitionColor = new TransitionColor(StaticColor.RED, StaticColor.GREEN, 3);
        expect(transitionColor.getColor()).toEqual(StaticColor.RED);
        expect(transitionColor.transitionFinished()).toBe(false);
        const midColor = transitionColor.getColor();
        expect(midColor.red).toBe(127);
        expect(midColor.green).toBe(127);
        expect(midColor.blue).toBe(0);
        expect(transitionColor.transitionFinished()).toBe(false);
        expect(transitionColor.getColor()).toEqual(StaticColor.GREEN);
        expect(transitionColor.transitionFinished()).toBe(true);
    });

    test('testSameColorTransition', () => {
        const transitionColor = new TransitionColor(StaticColor.YELLOW, StaticColor.YELLOW, 3);
        expect(transitionColor.getColor()).toEqual(StaticColor.YELLOW);
        expect(transitionColor.transitionFinished()).toBe(false);
        expect(transitionColor.getColor()).toEqual(StaticColor.YELLOW);
        expect(transitionColor.transitionFinished()).toBe(false);
        expect(transitionColor.getColor()).toEqual(StaticColor.YELLOW);
        expect(transitionColor.transitionFinished()).toBe(true);
    });
});
